README.md

```diff
@@ -1 +1,264 @@
+LD_PRELOAD feature provided by the system loader.
+
+All syscalls issued by libc are intercepted. Syscalls made
+by code outside libc are not intercepted. In order to
+be able to issue syscalls that are not intercepted, a
+convenience function is provided by the library:
+```c
+long syscall_no_intercept(long syscall_number, ...);
+```
+
+Three environment variables control the operation of the library:
+
+*INTERCEPT_LOG* -- when set, the library logs each syscall intercepted
+to a file. If it ends with "-" the path of the file is formed by appending
+a process id to the value provided in the environment variable.
+E.g.: initializing the library in a process with pid 123 when the
+INTERCEPT_LOG is set to "intercept.log-" will result in a log file named
+intercept.log-123.
+
+*INTERCEPT_LOG_TRUNC -- when set to 0, the log file from INTERCEPT_LOG
+is not truncated.
+
+*INTERCEPT_HOOK_CMDLINE_FILTER* -- when set, the library
+checks the contents of the /proc/self/cmdline file.
+Hotpatching, and syscall intercepting is only done, if the
+last component of the first zero terminated string in
+/proc/self/cmdline matches the string provided
+in the environment variable. This can also be queried
+by the user of the library:
+```c
+int libc_hook_in_process_allowed(void);
+```
+
+##### Example: #####
+
+```c
+#include <libsyscall_intercept_hook_point.h>
+#include <syscall.h>
+#include <errno.h>
+
+static int
+hook(long syscall_number,
+			long arg0, long arg1,
+			long arg2, long arg3,
+			long arg4, long arg5,
+			long *result)
+{
+	if (syscall_number == SYS_getdents) {
+		/*
+		 * Prevent the application from
+		 * using the getdents syscall. From
+		 * the point of view of the calling
+		 * process, it is as if the kernel
+		 * would return the ENOTSUP error
+		 * code from the syscall.
+		 */
+		*result = -ENOTSUP;
+		return 0;
+	} else {
+		/*
+		 * Ignore any other syscalls
+		 * i.e.: pass them on to the kernel
+		 * as would normally happen.
+		 */
+		return 1;
+	}
+}
+
+static __attribute__((constructor)) void
+init(void)
+{
+	// Set up the callback function
+	intercept_hook_point = hook;
+}
+```
+
+```sh
+$ cc example.c -lsyscall_intercept -fpic -shared -o example.so
+$ LD_LIBRARY_PATH=. LD_PRELOAD=example.so ls
+ls: reading directory '.': Operation not supported
+```
+
+# Under the hood: #
+
+##### Assumptions: #####
+In order to handle syscalls in user space, the library relies
+on the following assumptions:
+
+- Each syscall made by the applicaton is issued via libc
+- No other facility attempts to hotpatch libc in the same process
+- The libc implementation is already loaded in the processes
+memory space when the intercepting library is being initialized
+- The machine code in the libc implementation is suitable
+for the methods listed in this section
+- For some more basic assumptions, see the section on limitations.
+
+##### Disassembly: #####
+The library disassembles the text segment of the libc loaded
```

Comment:
Well, only text is interpreted as machine code, capstone is used to disassemble that. The other two are parsed to look for symbols, that can be mentioned, yes.